crates/subspace-networking/src/peer_info.rs

```diff
@@ -2,28 +2,54 @@ mod handler;
 mod protocol;
 
 use crate::peer_info::handler::HandlerInEvent;
+use event_listener_primitives::HandlerId;
 use handler::Handler;
 pub use handler::{Config, PeerInfoError, PeerInfoSuccess};
 use libp2p::core::{Endpoint, Multiaddr};
 use libp2p::swarm::behaviour::{ConnectionEstablished, FromSwarm};
 use libp2p::swarm::{
-    ConnectionDenied, ConnectionId, NetworkBehaviour, NotifyHandler, PollParameters, THandler,
-    THandlerInEvent, THandlerOutEvent, ToSwarm,
+    ConnectionClosed, ConnectionDenied, ConnectionId, NetworkBehaviour, NotifyHandler,
+    PollParameters, THandler, THandlerInEvent, THandlerOutEvent, ToSwarm,
 };
 use libp2p::PeerId;
 use parity_scale_codec::{Decode, Encode};
-use std::collections::VecDeque;
+use std::collections::{HashSet, VecDeque};
+use std::fmt;
+use std::sync::atomic::{AtomicBool, Ordering};
+use std::sync::Arc;
 use std::task::{Context, Poll};
 use tracing::debug;
 
-#[derive(Clone, Debug, Encode, Decode)]
+#[derive(Debug, Clone, Copy)]
+/// Peer info notification marker.
+pub struct Notification;
```

Comment:
Why not `()` or even `PeerInfo` rather than require catching useless `Notification` and then calling `.peer_info()` anyway (else why did you subscribe?).

Reply:
But you don't care about `Notification`, you care about `PeerInfo`. Why not sending notification with `PeerInfo` instead?